DEPR: Deprecate using `xlrd` engine for read_excel

roberthdevries opened the pull request:
- [x] closes #28547
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This is MR #29375 but rebased to master

cruzzoe committed:
Deprecate using `xlrd` engine and change default engine to read excel files to openpyxl

MarcoGorelli reviewed (changes requested):
Thanks @roberthdevries 

Can you use the method described here https://pandas.pydata.org/pandas-docs/stable/development/contributing.html#testing-warnings to test for warnings?

st-pasha commented on pandas/io/excel/_base.py:
```diff
@@ -852,6 +853,14 @@ def __init__(self, path_or_buffer, engine=None):
                 ext = os.path.splitext(str(path_or_buffer))[-1]
                 if ext == ".ods":
                     engine = "odf"
+
```
```suggestion
if ext == ".xls":
    engine = "xlrd"
```
Given that openpyxl doesn't support old `.xls` files.

roberthdevries replied:
Added

st-pasha commented on pandas/io/excel/_base.py:
```diff
@@ -852,6 +853,14 @@ def __init__(self, path_or_buffer, engine=None):
                 if ext == ".ods":
                     engine = "odf"
+
+        if engine == "xlrd":
```
The warning should not be issued when the parameter `engine="xlrd"` is passed explicitly.

roberthdevries replied:
Hmm, if the engine is deprecated, I would expect that all uses should be discouraged. Explicit or implicit.

WillAyd replied:
xlrd is the only thing that will read legacy .xls files unfortunately, so I don't think we need to outright remove all usage of it but want the default to switch to openpyxl

roberthdevries replied:
So are we already switching to openpyxl for everything other than .xls files (except of course .ods files and maybe .xlsb files)

WillAyd replied:
Yea xlsx and xlsm files (the former I would hope is what the vast majority of people read nowadays)

roberthdevries replied:
I have now changed the default engine to openpyxl and added a check to use xlrd for .xls files
This required quite some changes to the tests and a work-around for a rounding error in openpyxl.
See https://foss.heptapod.net/openpyxl/openpyxl/-/issues/1493

WillAyd commented on pandas/io/excel/_base.py:
```diff
@@ -844,14 +844,24 @@ class ExcelFile:
     def __init__(self, path_or_buffer, engine=None):
         if engine is None:
-            engine = "xlrd"
+            engine = "openpyxl"
```
So this actually changes the engine; I think the first step is to provide the warning that you have below that by default in the future we will switch to using openpyxl

WillAyd replied:
Making an exception for .xls files which have to use xlrd

roberthdevries replied:
So that means that a warning shall only be produced for .xlsx and .xlsm files that use xlrd?
And regarding the other remark about switching the engines, that was what you asked for a couple of comments back?

WillAyd replied:
That's right - warn first then change over time

roberthdevries replied:
so the change to use the openpyxl engine as the default has to be reverted?

roberthdevries replied:
@WillAyd Should I revert the change to make openpyxl the default engine?
And only warn when using xlrd in combination with .xlsx or .xlsm files?

WillAyd commented:
Yea should warn first can actually change in 2.0

Sent from my iPhone

> On Jul 3, 2020, at 1:25 PM, Robert de Vries <notifications@github.com> wrote:
> 
> ﻿
> @roberthdevries commented on this pull request.
> 
> In pandas/io/excel/_base.py:
> 
> > @@ -844,14 +844,24 @@ class ExcelFile:
>  
>      def __init__(self, path_or_buffer, engine=None):
>          if engine is None:
> -            engine = "xlrd"
> +            engine = "openpyxl"
> @WillAyd Should I revert the change to make openpyxl the default engine?
> And only warn when using xlrd in combination with .xlsx or .xlsm files?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

jreback commented:
@roberthdevries canyou update to comments and merge master

simonjayhawkins commented:
@roberthdevries can you address comments?

roberthdevries commented:
Not until I am back from vacation in three weeks.

roberthdevries committed:
Revert all changes related to switching to openpyxl as the default

simonjayhawkins commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -143,6 +143,7 @@ See :ref:`install.dependencies` and :ref:`install.optional_dependencies` for mor
 Deprecations
 ~~~~~~~~~~~~
 - Deprecated parameter ``inplace`` in :meth:`MultiIndex.set_codes` and :meth:`MultiIndex.set_levels` (:issue:`35626`)
+- :func:`read_excel` "xlrd" engine is deprecated for all file types that can be handled by "openpyxl" because "xlrd" is no longer maintained (:issue:`28547`).
```
can you reword for the benefit end users.

i.e. the default engine for `read_excel` is changing the the future

maybe say something like openpyxl is the recommended engine as xlrd is no longer maintained

roberthdevries replied:
done

simonjayhawkins commented on pandas/io/excel/_openpyxl.py:
```diff
@@ -517,7 +518,11 @@ def _convert_cell(self, cell, convert_float: bool) -> Scalar:
         # TODO: replace with openpyxl constants
         if cell.is_date:
-            return cell.value
+            try:
+                # workaround for inaccurate timestamp notation in excel
+                return datetime.fromtimestamp(round(cell.value.timestamp()))
+            except (AttributeError, OSError):
+                return cell.value
```
why is this changing?

roberthdevries replied:
This was a work-around for a bug in openpyxl (https://foss.heptapod.net/openpyxl/openpyxl/-/issues/1493), but is only apparent when you do a round trip save to xlsx and read back xlsx using openpyxl.
As this is not tested in any unit test, this can be removed. Agreed?

simonjayhawkins replied:
If adding code we should have a test. so either need to add test or can remove.

my preference would be to have this in a separate PR, so should raise pandas issue for this if removing.

roberthdevries replied:
Fair enough, remove it here and make a separate PR that includes a regression test.

roberthdevries committed:
Reword whatsnew message for the benefit of end users.

simonjayhawkins commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -144,6 +144,8 @@ Deprecations
 ~~~~~~~~~~~~
 - Deprecated parameter ``inplace`` in :meth:`MultiIndex.set_codes` and :meth:`MultiIndex.set_levels` (:issue:`35626`)
 - Deprecated parameter ``dtype`` in :~meth:`Index.copy` on method all index classes. Use the :meth:`Index.astype` method instead for changing dtype(:issue:`35853`)
+- :func:`read_excel` "xlrd" engine is deprecated. The recommended engine is "openpyxl" for "xlsx" and "xlsm" files, because "xlrd" is no longer maintained (:issue:`28547`).
```
thanks for updating. it's the ``read_excel` "xlrd" engine is deprecated` bit that I wanted removed

IIUC the xlrd is not deprecated. it's only that that default engine used will change.

bashtage replied:
Is it deprecated for use with xlsx files? IOW in the future, will only openpyxl be supported for xlsx?  Sounds reasonable to me.

simonjayhawkins replied:
hmm, that's not my understanding of the discussion in #28547

from https://github.com/pandas-dev/pandas/issues/28547#issue-496123767

> Considering that I think we need to deprecate using xlrd in favor of openpyxl. We might not necessarily need to remove the former and it does offer some functionality the latter doesn't (namely reading .xls files) but should at the very least start moving towards the latter

Indeed, the discussion has not explicitly mentioned disallowing `xlrd` for formats that `openpyxl` supports. but if the xlrd engine is not removed, we should decide now whether we would restrict it's use.

bashtage replied:
My read of this is to only keep xlrd for xls where it is required, and to deprecate where it is not.  In the long run, if xlrd breaks and no one takes over its maintenance, then we will either have to vendor xlrd or remove support for xls. In either case minimizing use of xlrd seems like a good idea to me.

roberthdevries replied:
@WillAyd is very specific regarding this issue. He states that we should warn first, then change the default (and maybe even remove the xlrd engine for xlsx and xlsm files altogether).
But not at this point, only a deprecation warning is asked, to notify users that this engine is no longer the preferred engine.

WillAyd replied:
Just to clarify we should only be changing the *default* reader to openpyxl. I think it's fine to keep xlrd around as a YMMV situation

roberthdevries replied:
Hmm, I just reverted the changes that made openpyxl the default. I am now very confused. I thought that this change was just about warning people about pending deprecation of the xlrd reader, not to switch them over already.

roberthdevries replied:
See your comments of July 1 and 3.

WillAyd replied:
I see where the wording is confusing, but yes we only warn now and change in the future. We always manage user-facing changes that way

WillAyd commented on pandas/io/excel/_base.py:
```diff
@@ -860,15 +861,26 @@ class ExcelFile:
+                if ext == "ods":
                     engine = "odf"
+
+        elif engine == "xlrd" and ext in ("xlsx", "xlsm"):
```
This warning should be in the `if engine is None` branch

roberthdevries replied:
Are you sure? This will mean that that people also get a warning when they ask for the default (which is still xlrd), instead of when they explicitly ask for xlrd.

WillAyd replied:
Yea so the point of it is that people who want to suppress the warnings will get a head start and explicitly request `engine="openpyxl"`, which is a good thing to sniff out any bugs

luk-f-a commented:
just because I didn't see it mentioned above, I thought I'd mention that `openpyxl` is 10x slower. So it's a good thing that the default is not being changed, because otherwise a lot of user would see an impact after upgrading.

erfannariman commented:
> just because I didn't see it mentioned above, I thought I'd mention that `openpyxl` is 10x slower. So it's a good thing that the default is not being changed, because otherwise a lot of user would see an impact after upgrading.

Interesting, what did you base this on?

luk-f-a commented:
> Interesting, what did you base this on?

Testing based on files I'm working on.

simonjayhawkins commented:
@roberthdevries This branch has conflicts that must be resolved. can you also address @WillAyd comments. https://github.com/pandas-dev/pandas/pull/35029#pullrequestreview-475475584

arw2019 commented:
Closing for now. @roberthdevries ping us whenever you'd like to continue and we'll reopen!

jreback commented:
@pandas-dev/pandas-core this would be good to do for 1.2, anyone interested in starting from here and putting up a PR?

rhshadrach commented:
I can finish this off, should be able to get a PR up by tomorrow morning.

rhshadrach committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into 35029

 Conflicts:
	doc/source/whatsnew/v1.2.0.rst
	pandas/io/excel/_base.py
	pandas/io/excel/_openpyxl.py
	pandas/tests/io/excel/test_readers.py
	pandas/tests/io/excel/test_xlrd.py

rhshadrach committed:
Fixed FutureWarning emitting logic, reverted openpyxl workaround

rhshadrach committed:
Revert change to stacklevel

rhshadrach commented:
A few notes that I wasn't 100% sure on:
 - I don't believe a doc page for ExcelFile gets created, but added the deprecation to the docstring there anyways.
 - For the future warning, stacklevel should be 2 for ExcelFile and 4 for pd.read_excel. I think 2 (and in general, the minimum) is correct here because that way there is an actual trace in all cases, but wasn't certain.
 - Added a deprecation note to the read_excel page:

![image](https://user-images.githubusercontent.com/45562402/100294481-eb843300-2f54-11eb-8ce5-8332f7910e47.png)

cc @jreback @simonjayhawkins @WillAyd

jreback commented on pandas/io/excel/_base.py:
```diff
@@ -878,12 +884,17 @@ class ExcelFile:
+        default ``xlrd`` for .xls* files, ``odf`` for .ods files.
         Engine compatibility :
-        - ``xlrd`` supports most old/new Excel file formats.
+        - ``xlrd`` supports most old/new Excel file formats but is no longer maintained.
```
i think can you add a note that xlrd not support for python >= 3.9 (everywhere you mention)

rhshadrach replied:
Sorry - I don't follow how this is different from python < 3.9, what do you mean by support? I am able to install xlrd on a python 3.9 environment:

![image](https://user-images.githubusercontent.com/45562402/100463250-86862580-3099-11eb-85ca-af4ef01000e4.png)

jreback replied:
see the issues when you actually use xlrd and pandas on 3.9. this is the entire reason for the deprecation; it is completely broken on 3.9

rhshadrach replied:
I'm not seeing that:

```
Python 3.9.0 (default, Nov 15 2020, 14:28:56) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pandas as pd
>>> pd.__version__
'1.2.0.dev0+1364.g65319af6e'
>>> pd.DataFrame({'a': [1, 2, 3]}).to_excel('test.xlsx')
>>> pd.read_excel('test.xlsx', engine='xlrd')
   Unnamed: 0  a
0           0  1
1           1  2
2           2  3
```

From the issue, I see this in the OP

> xlrd is unmaintained and the previous maintainer has asked us to move towards openpyxl. xlrd works now, but might have some issues when Python 3.9 or later gets released and changes some elements of the XML parser, as default usage right now throws a PendingDeprecationWarning

but no other mention of Python 3.9 being broken.

jreback replied:
#37795

jreback reviewed (changes requested):
a pre commit failure and some comments. ping on green.

jorisvandenbossche commented on pandas/io/excel/_base.py:
```diff
@@ -101,10 +102,15 @@
     If io is not a buffer or path, this must be set to identify io.
     Supported engines: "xlrd", "openpyxl", "odf", "pyxlsb", default "xlrd".
     Engine compatibility :
-    - "xlrd" supports most old/new Excel file formats.
```
Can you add a blank line above this one? (then the sphinx rendering will actually work)

jorisvandenbossche commented on pandas/io/excel/_base.py:
```diff
@@ -902,6 +913,12 @@ def __init__(
         self, path_or_buffer, engine=None, storage_options: StorageOptions = None
     ):
         if engine is None:
+            warnings.warn(
```
Shouldn't we be more specific about *when* to warn?

* If people only have openpyxl installed, shouldn't we default to use that instead of having the warning?
* From earlier discussion, I understood that we are (for now) fine to keep xlrd support for old excel files for which it is the only. So for those cases (eg `.xls`), we shouldn't raise a warning?

jreback replied:
and python < 3.9

rhshadrach replied:
@jorisvandenbossche I've implemented the xls check, however I don't understand your first bullet. Isn't having the deprecation warning so that we don't change behavior without notice? Why is it okay to change behavior if they have openpyxl installed?

jorisvandenbossche replied:
> Why is it okay to change behavior if they have openpyxl installed?

My suggestions was for when they have *only* openpyxl installed. It will currently not be a very common situation, since we don't try to use openpyxl if xlrd is not present, unless the user actually specifies `engine="openpyxl"`. But, given the deprecation warnings, this could be another solution for users: ensure you update your environment to only have openpyxl and not xlrd, and then you don't need to update all occurrences of `read_excel` in your code.

rhshadrach replied:
Ahh, thanks! I completely misread.

jorisvandenbossche commented on pandas/io/excel/_base.py:
```diff
@@ -902,6 +913,12 @@ def __init__(
         if engine is None:
+            warnings.warn(
+                "The default argument engine=None is deprecated. "
+                "Specify the engine argument to suppress this warning.",
```
I think we need to clarify this message, to explain the reason why it is deprecated, and what people can do: either specify "xlrd", or either install openpyxl and specify that

rhshadrach replied:
Changing to the following. Let me know of any thoughts.

> DeprecationWarning: The default argument engine=None is deprecated. Using None defaults to the xlrd engine which is no longer maintained. The default value will be 'openpyxl' in a future version of pandas, although xlrd will continue to be allowed for the indefinite future. Either install openpyxl and specify it as the engine, or specify 'xlrd' to suppress this warning.

jorisvandenbossche replied:
For the current behaviour, that sounds good!

jorisvandenbossche commented on pandas/io/excel/_base.py:
```diff
@@ -902,6 +913,12 @@ def __init__(
+                "The default argument engine=None is deprecated. "
+                "Specify the engine argument to suppress this warning.",
+                FutureWarning,
+                stacklevel=2,
```
> For the future warning, stacklevel should be 2 for ExcelFile and 4 for pd.read_excel. I think 2 (and in general, the minimum) is correct here because that way there is an actual trace in all cases, but wasn't certain.

Given that `read_excel` is used much more, it might make sense to ensure it is correct for `read_excel`

rhshadrach replied:
At the expense of having usage of ExcelFile directly have a stack trace that starts two levels too high (which includes the possibility of no stack trace)? IMO going too deep here is preferable, at least then the offending line is always visible.

Another solution would be using `inspect`:

```
import inspect
caller = inspect.stack()[1]
if (
    caller.filename.endswith("pandas/io/excel/_base.py")
    and caller.function == "read_excel"
):
    stacklevel = 4
else:
    stacklevel = 2
```

jreback replied:
if you want to do this is is fine, or ust turn off the check for stacklevel where needed. i don't think its worth jumping thru too many hoops.

jorisvandenbossche replied:
@rhshadrach as far as I know, warnings don't show a full stack trace, but only point to a single location. So whether this single location is wrong "higher" in the stack vs "lower" in the stack, I personally don't find the lower (deeper) one any more informative or less misleading .. 
The only advantage that the the deeper one has is that it will always show that the warning is coming from internal pandas code, so can give the hint this has to do with pandas functionality. But I think that goal can also be achieved by having a clear warning message (for example, include "pandas.read_excel" in the message)

jorisvandenbossche reviewed (changes requested):
There are several occurences in the docs that need to be fixed.

In general, I personally worry that this is going to be a quite annoying / noisy deprecation. `read_excel` is the second most used reader (after `read_csv`). And I might think that the majority of the users don't specify the engine argument. 
Do we know if there are differences in behaviour between openpyxl and xlrd for the resulting dataframe? Because if not, we could also already use openpyxl *if installed*, and only warn if xlrd is the only available engine.

Given how much this is used, it might also makes sense to start with a DeprecationgWarning, and only switch to the more noisy FutureWarning in a next release.

luk-f-a commented:
> > Because if not, we could also already use openpyxl _if installed_, and only warn if xlrd is the only available engine.
> 
> AFAIK openpyxl is slower than xlrd where they overlap, and so it might look like a perf issue to silently switch. Might have to be noisy about the default changing for a release or two to prompt users who care to specify a choice.

I can confirm the performance issue:

> After seeing the deprecation notice, we switched our code, and found out our CI was taking one hour longer (!). We traced it back to openpyxl.

jorisvandenbossche commented:
@luk-f-a with what size of files did you notice it? And what factor of slowdown?

I did a quick test locally with a file of 10,000 rows, 6 columns (5 floats and 1 string), and reading that with xlrd takes around 500ms, with openpyxl around 900ms (so indeed slower).

But so, it can indeed be slower. But even then, there is not yet really an alternative? (apart from keep relying on a dead project until is actually breaks)

Does somebody know if xlrd/openpyxl have APIs to get the size of an excel file / sheet? (eg possible idea could be to take a threshold for which we warn, and for small files where performance is not an issue, we could then still automatically use openpyxl when available)

luk-f-a commented:
@jorisvandenbossche I have a file with 38 rows and 2 columns (tiny!), taking 5.65s with openpyxl and 387ms with xlrd. Since it seems that openpyxl is impossible to avoid in the long term I'll open an issue with openpyxl. This difference is too large for such a simple file.

bashtage commented:
> I have a file with 38 rows and 2 columns (tiny!), taking 5.65s with openpyxl and 387ms with xlrd. Since it seems that openpyxl is impossible to avoid in the long term I'll open an issue with openpyxl. This difference is too large for such a simple file.

Is the file heavily formatted?  This is shockingly long.

luk-f-a commented:
at first sight it looks like a normal table. there must be something invisible because copying and pasting the content (including text format) into a new file reduces the time.

jorisvandenbossche commented:
Is it an excel file you can share? 
But indeed, that sounds as an issue in openpyxl, so would be good to report it there.

rhshadrach committed:
-

luk-f-a commented:
I was about to share it but then I found the reason for the slow reading performance: a link to another file. Even though no formula uses that link, and independently of whether that file exists or not (I tested on two different computers), it slows down reading the excel file. Unfortunately now I cannot share it because the link shows the entire internal file structure. I tried to reproduce by linking to another file (with a safe path) but the problem does not re-appear.
So maybe the performance problem I had is not so common, only with some kind of linked files.

rhshadrach committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into 35029

rhshadrach committed:
Changes from review
 - Added engine= throughout the docs
 - Warning suppressed for xls file or xlrd is not installed
 - Changed FutureWarning -> DeprecationWarning
 - Added "not supported for Python >= 3.9" note to docstrings and warning

rhshadrach commented:
cc @jreback @jorisvandenbossche

Thanks for the reviews, changes made. With the change from emitting a FutureWarning to now a DeprecationWarning, I don't see a way to check stacklevel in the tests because of the other DeprecationWarning. So even though the stacklevel is now properly set for both read_excel and ExcelFile, it isn't tested for.

In updating the docs,  `user_guide/io.rst` has some examples with xls files. I will open an issue to remove/convert those.

A few minor questions:
 - In adding `engine="openpyxl"` to read_excel calls in the docs to avoid the warnings, I even did the 0.17 whatsnew. I think that's okay?
 - To check if xlrd is installed, I used import_optional_dependency; is there a better way? Couldn't find any other checks for optional dependencies.
 - Should a note/explanation about the deprecation be added to `intro_tutorial/02_read_write.rst`, `user_guide/10min.rst`, and `user_guide/io.rst` where we need to add engine="openpyxl"? I can do so tomorrow.

jreback commented on pandas/io/excel/_base.py:
```diff
@@ -903,13 +919,51 @@ def __init__(
+                    "pandas, although xlrd will continue to be allowed for the "
+                    "indefinite future. Either install openpyxl and specify it as "
+                    "the engine or specify 'xlrd' to suppress this warning.",
+                    DeprecationWarning,
```
why are we not just raising a FutureWarning? sure this might throw lots of warnings in user code but that is exactly where this needs to be changed and *not* library code (where a DeprecationWarning would be useful)

jorisvandenbossche replied:
Fore sure read_excel will also be used in library code

jreback replied:
yes it will be used in library code
but the vast majority would be in use code

jreback commented:
> cc @jreback @jorisvandenbossche
> 
> Thanks for the reviews, changes made. With the change from emitting a FutureWarning to now a DeprecationWarning, I don't see a way to check stacklevel in the tests because of the other DeprecationWarning. So even though the stacklevel is now properly set for both read_excel and ExcelFile, it isn't tested for.
> 
> In updating the docs, `user_guide/io.rst` has some examples with xls files. I will open an issue to remove/convert those.
> 
> A few minor questions:
> 
> * In adding `engine="openpyxl"` to read_excel calls in the docs to avoid the warnings, I even did the 0.17 whatsnew. I think that's okay?
> * Should a note/explanation about the deprecation be added to `intro_tutorial/02_read_write.rst`, `user_guide/10min.rst`, and `user_guide/io.rst` where we need to add engine="openpyxl"? I can do so tomorrow.

also would add a big warning box in the io.rst for reading excel about this change. also *maybe* add in the top of the whatsnew as well (a warning box).

rhshadrach committed:
DeprecationWarning -> FutureWarning; added warning to io.rst/whatsnew

rhshadrach commented:
@jreback Changes made, added the warnings to both io.rst and whatsnew. Can remove from whatsnew if we don't want it.

![image](https://user-images.githubusercontent.com/45562402/100548005-ee727280-3237-11eb-8f3c-b0b23f96d532.png)

![image](https://user-images.githubusercontent.com/45562402/100548017-fb8f6180-3237-11eb-96d0-b3df61d0fb1f.png)

jreback commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -8,6 +8,16 @@ including other versions of pandas.
 
+.. warning::
+
+   The default argument ``engine=None`` to ``pd.read_excel`` is
```
great if we think it's overkill than can remove before the release 

but can't hurt having it here

rhshadrach committed:
"to suppress this warning." -> "to avoid raising a FutureWarning."

rhshadrach commented:
One last tweak in the warning boxes added to io.rst/whatsnew: the end "...to suppress this warning." was out of context and has been changed to "...to avoid raising a FutureWarning."

jreback commented:
ok will merge on green. we can always change this if major pushback in the rc

jorisvandenbossche commented on pandas/io/excel/_base.py:
```diff
@@ -101,10 +103,17 @@
     - "pyxlsb" supports Binary Excel files.
+
+    .. deprecated:: 1.2.0
+        The default value ``None`` is deprecated and will be changed to ``"openpyxl"``
```
The default of None is only deprecated when you're reading an xls/xlsx file, not when reading any of the other formats like ods. 
It might be worth explaining here (outside of the deprecated box) what the default engine is if not specified (eg also that openpyxl is automatically used if xlrd is not present for xlsx files)

jorisvandenbossche reviewed (changes requested):
@rhshadrach thanks for all the updates!

One more case where we shouldn't warn if when the user is passing an xlrd Book object (it's clear which engine you are using then).

There are also still plenty of warnings raised by the tests. We could maybe ignore then with pytest on a file or class level (with pytest.mark.filterwarnings)

And I want to repeat my question from above: do we know about differences in behaviour between both engines (apart from performance)? (to which no one answered, it's probably also not an easy question)  
Because if not, I think we should consider automatically using openpyxl when possible, and only raise a warning when it is not installed.

jreback commented:
> And I want to repeat my question from above: do we know about differences in behaviour between both engines (apart from performance)? (to which no one answered, it's probably also not an easy question)
Because if not, I think we should consider automatically using openpyxl when possible, and only raise a warning when it is not installed.

isn't this part of the point?

jreback commented:
> There are also still plenty of warnings raised by the tests. We could maybe ignore then with pytest on a file or class level (with pytest.mark.filterwarnings)

yeah should address this here

jreback commented:
@rhshadrach ping when this is updated.

jorisvandenbossche commented:
> isn't this part of the point?

Sorry, I don't understand what you are trying to say

jreback commented:
> > isn't this part of the point?
> 
> Sorry, I don't understand what you are trying to say

well i think the context was in my reply to you. openpyxl should be the default as its the supported library moving forward.

jorisvandenbossche commented:
> openpyxl should be the default as its the supported library moving forward.

That's indeed what I proposed, but that's not what this PR is doing

jreback commented:
> > openpyxl should be the default as its the supported library moving forward.
> 
> That's indeed what I proposed, but that's not what this PR is doing

this is a precursor to doing exactly this.

jorisvandenbossche commented:
And so my question is: why not do it directly?

jreback commented:
> And so my question is: why not do it directly?

umm, you mean change the default? well we would need to deprecate it *anyhow*, i don't see how

> And so my question is: why not do it directly?

is practically anything different here?

 (with the exception of for some cases we still recommend xlrd, e.g. older .xls files), which to be honest is probably fine.

jorisvandenbossche commented:
> you mean change the default? well we would need to deprecate it anyhow, i don't see how

Yes, or at least: change the order in which we check them: prefer openpyxl if both are installed (instead of the other way around as is the case now). 
Of course, if the user has *only* xlrd installed, we will still need to raise a deprecation warning.

> is practically anything different here?

That's exactly what I have been asking ;) 
I am not very familiar with the details of the different excel engines. But if they give the same resulting dataframe, we are annoying the user with a deprecation warning while the actual behaviour (the returned dataframe) might not actually change.

jreback commented:
> > you mean change the default? well we would need to deprecate it anyhow, i don't see how
> 
> Yes, or at least: change the order in which we check them: prefer openpyxl if both are installed (instead of the other way around as is the case now).
> Of course, if the user has _only_ xlrd installed, we will still need to raise a deprecation warning.
> 

This sounds reasonable to do.

> > is practically anything different here?
> 
> That's exactly what I have been asking ;)
> I am not very familiar with the details of the different excel engines. But if they give the same resulting dataframe, we are annoying the user with a deprecation warning while the actual behaviour (the returned dataframe) might not actually change.

Again we can just do it. We prefer a maintained to a non-maintened package, even at the expense of an immediate change.

rhshadrach commented:
So then we are no longer deprecating the default argument ``engine=None`` then, or changing it to "openpyxl" in the future, correct? If that's the case, I propose the new warning message when using ExcelFile (including via read_excel):

```
   FutureWarning: The xlrd engine is no
   longer maintained, and is not supported when using pandas with
   python >= 3.9, although xlrd will continue to be allowed for the
   indefinite future. Either install openpyxl or specify engine='xlrd' to silence this warning.
```

and for the whatsnew:

```
   Previously, the default argument ``engine=None`` to ``pd.read_excel``
   would result in using the xlrd engine in many cases. The xlrd engine is no
   longer maintained, and is not supported when using pandas with
   python >= 3.9. If ``engine="None"``, pandas will now use the following behavior:
    - If the file or stream is of ods type, then the ods engine will be used.
    - Otherwise if the file has the extension `.xls`, then xlrd will be used.
    - Otherwise if openpyxl is installed, then openpyxl will be used.
    - Otherwise xlrd will be used and a FutureWarning will be raised.
   Specifying ``engine="xlrd"`` will continue to be allowed for the indefinite future.
```

@jreback @jorisvandenbossche can you confirm I summarized the behavior correctly?

Since behavior is simply changing, it seems to me that we no longer need a warning in io.rst, and the addition of `engine=` to the docs can be removed.

rhshadrach commented:
Attempting to make the changes, the test `io.excel.test_readers.TestReaders.test_date_conversion_overflow` fails with

> OverflowError: signed integer is greater than maximum

when using openpyxl on an xlsx file (where we used to use xlrd). I think this answers @jorisvandenbossche question in the negative: there are differences between the frames read in by xlrd and openpyxl.

jreback commented:
ok so the question here is should we change the default if we get identical results (i think answer is a clear yes)

but now we see an i supported case

i think this is prob not a blocker (and in fact we could maybe make a better error message for this case that says you should just use xlrd)

rhshadrach commented:
Looking at the tests further, I didn't notice they were failing to run due to `test_readers.test_excel_high_surrogate`. On the line `actual = pd.read_excel("high_surrogate.xlsx")` I get:

> INTERNALERROR> UnicodeEncodeError: 'utf-8' codec can't encode character '\udc88' in position 1041: surrogates not allowed

and this makes the rest of the tests not run. At least 15 are failing. I'll have another few hours to look into this tonight, but otherwise I won't be available until 16:00 EST (21:00 UTC)

rhshadrach committed:
Changed engine=None to mostly using openpyxl

rhshadrach committed:
Minor doc touchups

rhshadrach commented:
@jreback @jorisvandenbossche 

Changes made. I've adjusted the tests only as necessary to either (i) test for openpyxl as being the default or (ii) specified engine="xlrd" if openpyxl would fail and either it seemed the test was specifically for xlrd or there was no clear way to resolve the test otherwise.

I don't know how to write a test for the FutureWarning - in testing environment, I think openpyxl will always be installed and so it will never be raised. As such, I've simply removed the tests for the FutureWarning for now.

As mentioned above, I won't be available until 16:00 EST (21:00 UTC). Anyone is of course welcomed to push this over the finish line.

jreback commented:
we don't have either installed in the numpy dev environment so can certainly add some basic tests that run (likely we skip everything if we don't have either installed)

rhshadrach committed:
Re-added tests, minor doc touchups

rhshadrach committed:
Test for no warning as well

rhshadrach commented:
Linux py37_minimum_versions failure is `pandas/tests/resample/test_deprecated.py:98`, unrelated.
macOS py37 failed to start, trying again. 
All other builds passed.

rhshadrach commented:
/azp run

rhshadrach commented:
@jreback tests pass; two failures below on Linux py38_np_dev are unrelated and passed on the previous run.

```
FAILED pandas/tests/series/indexing/test_indexing.py::test_loc_setitem_2d_to_1d_raises
FAILED pandas/tests/util/test_show_versions.py::test_show_versions - Assertio...
```

simonjayhawkins commented:
> @jreback tests pass; two failures below on Linux py38_np_dev are unrelated and passed on the previous run.
> 
> ```
> FAILED pandas/tests/series/indexing/test_indexing.py::test_loc_setitem_2d_to_1d_raises
> FAILED pandas/tests/util/test_show_versions.py::test_show_versions - Assertio...
> ```

these tests are failing on other PRs too.

jreback commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -8,6 +8,20 @@ including other versions of pandas.
+.. warning::
+
+   Previously, the default argument ``engine=None`` to ``pd.read_excel``
+   would result in using the xlrd engine in many cases. The engine xlrd is no longer
```
double back-tick on xlrd (alt can put a link to xlrd itself, e.g. https://xlrd.readthedocs.io/en/latest/)

jreback commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -8,6 +8,20 @@ including other versions of pandas.
+   following logic is now used to determine the engine.
+
+   - If ``path_or_buffer`` is an OpenDocument format (.odf, .ods, .odt), then odf will be used.
+   - Otherwise if ``path_or_buffer`` is a bytes stream, the file has the extension ``.xls``, or is an xlrd Book instance, then xlrd will be used.
```
double backtick xlrd / odf (only put the docs link on L14)

jreback commented on pandas/io/excel/_base.py:
```diff
@@ -99,12 +102,29 @@
     - "pyxlsb" supports Binary Excel files.
+
+    .. versionchanged:: 1.2.0
+        The engine xlrd is no longer maintained, and is not supported with
```
I *think* need a blank line here to render (make this section the same as in the whatsnew as per formatting)

jorisvandenbossche replied:
No, for versionchanged this is OK (rst .. ;-))

jreback commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -8,6 +8,20 @@ including other versions of pandas.
+   maintained, and is not supported with python >= 3.9. When ``engine=None``, the
+   following logic is now used to determine the engine.
+
+   - If ``path_or_buffer`` is an OpenDocument format (.odf, .ods, .odt), then odf will be used.
```
link for odf: https://pypi.org/project/odfpy/

jreback commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -8,6 +8,20 @@ including other versions of pandas.
+
+   - If ``path_or_buffer`` is an OpenDocument format (.odf, .ods, .odt), then odf will be used.
+   - Otherwise if ``path_or_buffer`` is a bytes stream, the file has the extension ``.xls``, or is an xlrd Book instance, then xlrd will be used.
+   - Otherwise if openpyxl is installed, then openpyxl will be used.
```
link: https://pypi.org/project/openpyxl/

jreback commented on pandas/io/excel/_base.py:
```diff
@@ -877,13 +897,29 @@ class ExcelFile:
+           python >= 3.9. When ``engine=None``, the following logic will be
+           used to determine the engine.
+
+           - If ``path_or_buffer`` is an OpenDocument format (.odf, .ods, .odt),
```
obviously as much of the formatting you can do here as well

jreback reviewed (commented):
just some formatting considerations. ping when pushed as ok for merge. cc @jorisvandenbossche

jorisvandenbossche commented on doc/source/whatsnew/v1.2.0.rst:
```diff
@@ -8,6 +8,20 @@ including other versions of pandas.
+   maintained, and is not supported with python >= 3.9. When ``engine=None``, the
+   following logic is now used to determine the engine.
+
+   - If ``path_or_buffer`` is an OpenDocument format (.odf, .ods, .odt), then odf will be used.
+   - Otherwise if ``path_or_buffer`` is a bytes stream, the file has the extension ``.xls``, or is an xlrd Book instance, then xlrd will be used.
+   - Otherwise if openpyxl is installed, then openpyxl will be used.
+   - Otherwise xlrd will be used and a ``FutureWarning`` will be raised.
```
I would maybe rearrange this list: the most important piece of information we want to convey here is that for xlsx files the default changed from xlrd to openpyxl, if installed. So I would also put that on top of the list (or keep it just to this for the whatsnew, as the other items didn't change. The full list is still in the actual docs).

rhshadrach replied:
We don't actually look at the extension or the file format when determining the engine in various cases, so it isn't just changing for xlsx files, right? What do you think of this:

```
Previously, the default argument ``engine=None`` to ``pd.read_excel``
would result in using the `xlrd <https://xlrd.readthedocs.io/en/latest/>`_ engine
in many cases. The engine ``xlrd`` is no longer
maintained, and is not supported with python >= 3.9. If 
`openpyxl <https://pypi.org/project/openpyxl/>`_  is installed, many of these 
cases will now default to using the ``openpyxl`` engine. See the
:func:`read_excel` docs for more details.
```

jorisvandenbossche reviewed (commented):
Small comment on the whatsnew, but it's perfectly fine to only address this later after the RC as well, it's not a blocker

jreback commented:
> Small comment on the whatsnew, but it's perfectly fine to only address this later after the RC as well, it's not a blocker

ideally we do this now, because these docs are important (and small change)

rhshadrach committed:
Doc tweaks

jreback commented:
thanks @rhshadrach and @roberthdevries for this, very nice!

we may need some tweeks during the rc but can be later

rhshadrach commented:
Thanks @jreback. Happy to support if issues arise.
